## E-commerce Data cleaning 

#### The process of cleaning the data involved utilizing SQL, Python, Jupyter Notebook, and Excel to successfully complete this challenging task. Let's start by examining how it began in Python before transitioning to other programming languages.




In [1]:
#importing necessary libraries
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns
sns.set(style='dark')
import warnings 
warnings.filterwarnings('ignore')

In [20]:
#load datset
df = pd.read_csv('data.csv', index_col=4, header=0, infer_datetime_format=True, parse_dates=True, encoding='latin-1')

In [21]:
#View dataset
df

InvoiceNo StockCode                          Description  \
InvoiceDate                                                                    
2010-12-01 08:26:00    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER   
2010-12-01 08:26:00    536365     71053                  WHITE METAL LANTERN   
2010-12-01 08:26:00    536365    84406B       CREAM CUPID HEARTS COAT HANGER   
2010-12-01 08:26:00    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE   
2010-12-01 08:26:00    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.   
...                       ...       ...                                  ...   
2011-12-09 12:50:00    581587     22613          PACK OF 20 SPACEBOY NAPKINS   
2011-12-09 12:50:00    581587     22899         CHILDREN'S APRON DOLLY GIRL    
2011-12-09 12:50:00    581587     23254        CHILDRENS CUTLERY DOLLY GIRL    
2011-12-09 12:50:00    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE   
2011-12-09 12:50:00    581587     22138        BAKING SET 9 PIECE RETROSPOT    

                     Quantity  UnitPrice  CustomerID         Country  
InvoiceDate                                                           
2010-12-01 08:26:00         6       2.55     17850.0  United Kingdom  
2010-12-01 08:26:00         6       3.39     17850.0  United Kingdom  
2010-12-01 08:26:00         8       2.75     17850.0  United Kingdom  
2010-12-01 08:26:00         6       3.39     17850.0  United Kingdom  
2010-12-01 08:26:00         6       3.39     17850.0  United Kingdom  
...                       ...        ...         ...             ...  
2011-12-09 12:50:00        12       0.85     12680.0          France  
2011-12-09 12:50:00         6       2.10     12680.0          France  
2011-12-09 12:50:00         4       4.15     12680.0          France  
2011-12-09 12:50:00         4       4.15     12680.0          France  
2011-12-09 12:50:00         3       4.95     12680.0          France  

[541909 rows x 7 columns]

#### The dataset contains so many things that will hinder effective analysis. The invoice number and stock code aren't very necessary in this analysis. Also the unit price and quantity must give us a total. All this will be done.

In [22]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'UnitPrice',
       'CustomerID', 'Country'],
      dtype='object')

In [24]:
#checking for null values in the dataset set
df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

#### Dropping the StockNo, InvoiceNo and CustomerID columns

In [31]:
df.drop('InvoiceNo', axis=1, inplace=True)

In [32]:
df.drop('StockCode', axis=1, inplace=True)

In [36]:
df.drop('CustomerID', axis=1, inplace=True)

In [37]:
df

Description  Quantity  UnitPrice  \
InvoiceDate                                                                     
2010-12-01 08:26:00   WHITE HANGING HEART T-LIGHT HOLDER         6       2.55   
2010-12-01 08:26:00                  WHITE METAL LANTERN         6       3.39   
2010-12-01 08:26:00       CREAM CUPID HEARTS COAT HANGER         8       2.75   
2010-12-01 08:26:00  KNITTED UNION FLAG HOT WATER BOTTLE         6       3.39   
2010-12-01 08:26:00       RED WOOLLY HOTTIE WHITE HEART.         6       3.39   
...                                                  ...       ...        ...   
2011-12-09 12:50:00          PACK OF 20 SPACEBOY NAPKINS        12       0.85   
2011-12-09 12:50:00         CHILDREN'S APRON DOLLY GIRL          6       2.10   
2011-12-09 12:50:00        CHILDRENS CUTLERY DOLLY GIRL          4       4.15   
2011-12-09 12:50:00      CHILDRENS CUTLERY CIRCUS PARADE         4       4.15   
2011-12-09 12:50:00        BAKING SET 9 PIECE RETROSPOT          3       4.95   

                            Country  
InvoiceDate                          
2010-12-01 08:26:00  United Kingdom  
2010-12-01 08:26:00  United Kingdom  
2010-12-01 08:26:00  United Kingdom  
2010-12-01 08:26:00  United Kingdom  
2010-12-01 08:26:00  United Kingdom  
...                             ...  
2011-12-09 12:50:00          France  
2011-12-09 12:50:00          France  
2011-12-09 12:50:00          France  
2011-12-09 12:50:00          France  
2011-12-09 12:50:00          France  

[540415 rows x 4 columns]

#### Let's change the data types of the columns

In [38]:
df.dtypes

Description     object
Quantity         int64
UnitPrice      float64
Country         object
dtype: object

#### We're all set to proceed! Now, let's proceed with creating a column that will store the product of the quantity and unit price of the items. Additionally, we'll create another column to categorize the items using SQL.




In [40]:
df['Total'] = df['Quantity'] * df['UnitPrice']
df

Description  Quantity  UnitPrice  \
InvoiceDate                                                                     
2010-12-01 08:26:00   WHITE HANGING HEART T-LIGHT HOLDER         6       2.55   
2010-12-01 08:26:00                  WHITE METAL LANTERN         6       3.39   
2010-12-01 08:26:00       CREAM CUPID HEARTS COAT HANGER         8       2.75   
2010-12-01 08:26:00  KNITTED UNION FLAG HOT WATER BOTTLE         6       3.39   
2010-12-01 08:26:00       RED WOOLLY HOTTIE WHITE HEART.         6       3.39   
...                                                  ...       ...        ...   
2011-12-09 12:50:00          PACK OF 20 SPACEBOY NAPKINS        12       0.85   
2011-12-09 12:50:00         CHILDREN'S APRON DOLLY GIRL          6       2.10   
2011-12-09 12:50:00        CHILDRENS CUTLERY DOLLY GIRL          4       4.15   
2011-12-09 12:50:00      CHILDRENS CUTLERY CIRCUS PARADE         4       4.15   
2011-12-09 12:50:00        BAKING SET 9 PIECE RETROSPOT          3       4.95   

                            Country  Total  
InvoiceDate                                 
2010-12-01 08:26:00  United Kingdom  15.30  
2010-12-01 08:26:00  United Kingdom  20.34  
2010-12-01 08:26:00  United Kingdom  22.00  
2010-12-01 08:26:00  United Kingdom  20.34  
2010-12-01 08:26:00  United Kingdom  20.34  
...                             ...    ...  
2011-12-09 12:50:00          France  10.20  
2011-12-09 12:50:00          France  12.60  
2011-12-09 12:50:00          France  16.60  
2011-12-09 12:50:00          France  16.60  
2011-12-09 12:50:00          France  14.85  

[540415 rows x 5 columns]

In [41]:
df['Categories'] = df['Description']

In [42]:
df

Description  Quantity  UnitPrice  \
InvoiceDate                                                                     
2010-12-01 08:26:00   WHITE HANGING HEART T-LIGHT HOLDER         6       2.55   
2010-12-01 08:26:00                  WHITE METAL LANTERN         6       3.39   
2010-12-01 08:26:00       CREAM CUPID HEARTS COAT HANGER         8       2.75   
2010-12-01 08:26:00  KNITTED UNION FLAG HOT WATER BOTTLE         6       3.39   
2010-12-01 08:26:00       RED WOOLLY HOTTIE WHITE HEART.         6       3.39   
...                                                  ...       ...        ...   
2011-12-09 12:50:00          PACK OF 20 SPACEBOY NAPKINS        12       0.85   
2011-12-09 12:50:00         CHILDREN'S APRON DOLLY GIRL          6       2.10   
2011-12-09 12:50:00        CHILDRENS CUTLERY DOLLY GIRL          4       4.15   
2011-12-09 12:50:00      CHILDRENS CUTLERY CIRCUS PARADE         4       4.15   
2011-12-09 12:50:00        BAKING SET 9 PIECE RETROSPOT          3       4.95   

                            Country  Total  \
InvoiceDate                                  
2010-12-01 08:26:00  United Kingdom  15.30   
2010-12-01 08:26:00  United Kingdom  20.34   
2010-12-01 08:26:00  United Kingdom  22.00   
2010-12-01 08:26:00  United Kingdom  20.34   
2010-12-01 08:26:00  United Kingdom  20.34   
...                             ...    ...   
2011-12-09 12:50:00          France  10.20   
2011-12-09 12:50:00          France  12.60   
2011-12-09 12:50:00          France  16.60   
2011-12-09 12:50:00          France  16.60   
2011-12-09 12:50:00          France  14.85   

                                              Categories  
InvoiceDate                                               
2010-12-01 08:26:00   WHITE HANGING HEART T-LIGHT HOLDER  
2010-12-01 08:26:00                  WHITE METAL LANTERN  
2010-12-01 08:26:00       CREAM CUPID HEARTS COAT HANGER  
2010-12-01 08:26:00  KNITTED UNION FLAG HOT WATER BOTTLE  
2010-12-01 08:26:00       RED WOOLLY HOTTIE WHITE HEART.  
...                                                  ...  
2011-12-09 12:50:00          PACK OF 20 SPACEBOY NAPKINS  
2011-12-09 12:50:00         CHILDREN'S APRON DOLLY GIRL   
2011-12-09 12:50:00        CHILDRENS CUTLERY DOLLY GIRL   
2011-12-09 12:50:00      CHILDRENS CUTLERY CIRCUS PARADE  
2011-12-09 12:50:00        BAKING SET 9 PIECE RETROSPOT   

[540415 rows x 6 columns]

#### This is what the dataset looks like now. Let's jump to SQL.

~~~SQL
--creating subcategories from the Categories column

UPDATE sql_ecommerce
SET Categories = 'STATIONARY'
WHERE Categories LIKE '%INFLATABLE POLITICAL GLOBE %';

~~~SQL
UPDATE sql_ecommerce
SET Categories = 'HOME AND OFFICE DECOR'
WHERE Categories = 'WHITE HANGING HEART T-LIGHT HOLDER';

UPDATE sql_ecommerce
SET Categories = 'HOME AND OFFICE DECOR'
WHERE Categories = 'WHITE METAL LANTERN';

~~~SQL
UPDATE sql_ecommerce
SET Categories = 'HOME AND OFFICE DECOR'
WHERE Categories LIKE '%HAND WARMER RED POLKA DOT%';


UPDATE sql_ecommerce
SET Categories = 'PERSONAL'
WHERE Categories LIKE '%HAND WARMER UNION JACK%';


UPDATE sql_ecommerce
SET Categories = 'GIFT'
WHERE Categories LIKE '%NINJA RABBIT BLACK%';

~~~SQL
UPDATE sql_ecommerce
SET Categories = 'PERSONAL'
WHERE Categories LIKE '%OCEAN STRIPE HAMMOCK %';


UPDATE sql_ecommerce
SET Categories = 'JEWELRY'
WHERE Categories LIKE '%LARIAT%'
					   

UPDATE sql_ecommerce
SET Categories = 'GIFTS'
WHERE Categories LIKE '%DOUGHNUT LIP GLOSS %'

					   
UPDATE sql_ecommerce
SET Categories = 'HOME AND OFFICE DECOR'
WHERE Categories LIKE (
					   '%DISCO BALL ROTATOR BATTERY OPERATED%'
					   )

~~~SQL
UPDATE sql_ecommerce
SET Categories = 'FELT TOADSTOOL SMALL'
WHERE Categories = 'FELT TOADSTOOL  SMALL';



UPDATE sql_ecommerce
SET Categories = NULL
WHERE Categories = 'GLOW IN DARK DOLPHINS';

--Multiple items were removed but not all were recorded

~~~SQL
--Removing null values
DELETE FROM sql_ecommerce
WHERE Categories IS NULL;

UPDATE sql_ecommerce
set Country = UPPER(Country)

UPDATE sql_ecommerce
set Country = 'IRELAND'
where Country = 'EIRE'


UPDATE sql_ecommerce
set Country = 'SOUTH AFRICA'
where Country = 'RSA'

--SELECT *
--FROM sql_ecommerce
--WHERE Country = 'EUROPEAN COMMUNITY'
--ORDER BY 1 ASC


#### This is what categorizing the items in the Categories column looks like. This provided a column which has a category for each item in the Description column.

~~~SQL
SELECT DISTINCT Categories
FROM sql_ecommerce
ORDER BY Categories DESC

| Categories             |
|-----------------------|
| STATIONARY            |
| SIGNS                 |
| PERSONAL              |
| JEWELRY               |
| HOME AND OFFICE DECOR |
| GIFTS                 |


#### After all the work put into it, a table which is fit for analysis was obtained

~~~SQL
SELECT TOP 10*
FROM sql_ecommercerce

|     InvoiceDate    |          Description         | Quantity | UnitPrice |    Country     | Total |       Categories       |
|-----------------------|-----------------------------|----------|-----------|----------------|--------|------------------------|
| 2010-12-01 08:26:00.000 | WHITE HANGING HEART T-LIGHT HOLDER |    6    |    2.55   | UNITED KINGDOM |  15.3  | HOME AND OFFICE DECOR  |
| 2010-12-01 08:26:00.000 |         WHITE METAL LANTERN       |    6    |    3.39   | UNITED KINGDOM |  20.3  | HOME AND OFFICE DECOR  |
| 2010-12-01 08:26:00.000 |  CREAM CUPID HEARTS COAT HANGER   |    8    |    2.75   | UNITED KINGDOM |   22   | HOME AND OFFICE DECOR  |
| 2010-12-01 08:26:00.000 | KNITTED UNION FLAG HOT WATER BOTTLE |    6    |    3.39   | UNITED KINGDOM |  20.3  |       PERSONAL         |
| 2010-12-01 08:26:00.000 |   RED WOOLLY HOTTIE WHITE HEART   |    6    |    3.39   | UNITED KINGDOM |  20.3  | HOME AND OFFICE DECOR  |
| 2010-12-01 08:26:00.000 |     SET 7 BABUSHKA NESTING BOXES   |    2    |    7.65   | UNITED KINGDOM |  15.3  |         GIFTS          |
| 2010-12-01 08:26:00.000 | GLASS STAR FROSTED T-LIGHT HOLDER |    6    |    4.25   | UNITED KINGDOM |  25.5  | HOME AND OFFICE DECOR  |
| 2010-12-01 08:28:00.000 |       HAND WARMER UNION JACK      |    6    |    1.85   | UNITED KINGDOM |  11.1  |       PERSONAL         |
| 2010-12-01 08:28:00.000 |     HAND WARMER RED POLKA DOT     |    6    |    1.85   | UNITED KINGDOM |  11.1  | HOME AND OFFICE DECOR  |
| 2010-12-01 08:34:00.000 |   ASSORTED COLOUR BIRD ORNAMENT   |    32   |    1.69   | UNITED KINGDOM |  54.1  |        JEWELRY         |


#### After rearranging the columns in excel, this is what the dataset looks looks like;

In [43]:
df = pd.read_csv('categorized_ecommerce.csv', index_col=0, header=0, parse_dates=True, infer_datetime_format=True)

In [44]:
df

Description             Categories  \
InvoiceDate                                                               
2010-12-01    WHITE HANGING HEART T-LIGHT HOLDER  HOME AND OFFICE DECOR   
2010-12-01                   WHITE METAL LANTERN  HOME AND OFFICE DECOR   
2010-12-01        CREAM CUPID HEARTS COAT HANGER  HOME AND OFFICE DECOR   
2010-12-01   KNITTED UNION FLAG HOT WATER BOTTLE               PERSONAL   
2010-12-01        RED WOOLLY HOTTIE WHITE HEART.  HOME AND OFFICE DECOR   
...                                          ...                    ...   
2011-03-21                REVOLVER WOODEN RULER                   GIFTS   
2011-03-21             RED RETROSPOT OVEN GLOVE   HOME AND OFFICE DECOR   
2011-03-21   SET OF 2 TEA TOWELS APPLE AND PEARS                  GIFTS   
2011-03-21          STRAWBERRY FAIRY CAKE TEAPOT  HOME AND OFFICE DECOR   
2011-03-21                RED SPOTTY BISCUIT TIN  HOME AND OFFICE DECOR   

             Quantity  UnitPrice  Total         Country  
InvoiceDate                                              
2010-12-01          6       2.55   15.3  UNITED KINGDOM  
2010-12-01          6       3.39   20.3  UNITED KINGDOM  
2010-12-01          8       2.75   22.0  UNITED KINGDOM  
2010-12-01          6       3.39   20.3  UNITED KINGDOM  
2010-12-01          6       3.39   20.3  UNITED KINGDOM  
...               ...        ...    ...             ...  
2011-03-21          1       1.95    2.0          FRANCE  
2011-03-21          2       1.25    2.5          FRANCE  
2011-03-21          1       2.95    3.0          FRANCE  
2011-03-21          2       4.95    9.9          FRANCE  
2011-03-21          6       3.75   22.5          FRANCE  

[537186 rows x 6 columns]

In [49]:
df['Categories'].unique()

array(['HOME AND OFFICE DECOR', 'PERSONAL', 'GIFTS', 'JEWELRY',
       'STATIONARY', 'SIGNS'], dtype=object)

In [53]:
df.isnull().sum()

Description    0
Categories     0
Quantity       0
UnitPrice      0
Total          0
Country        0
dtype: int64